core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1077 @@
+        intermediaryOutputs.add(outputQueue);
+        QueuePusher<ResultBatch<T>> pusher = new QueuePusher<>(outputQueue, hasTimeout, timeoutAt);
+
+        List<BatchedResultsCursor<T>> partitionCursors = new ArrayList<>(sequences.size());
+        for (Sequence<T> s : partition) {
+          partitionCursors.add(new YielderBatchedResultsCursor<>(new SequenceBatcher<>(s, batchSize), orderingFn));
+        }
+        PrepareMergeCombineInputsAction<T> blockForInputsAction = new PrepareMergeCombineInputsAction<>(
+            partitionCursors,
+            pusher,
+            orderingFn,
+            combineFn,
+            yieldAfter,
+            batchSize,
+            cancellationGizmo
+        );
+        tasks.add(blockForInputsAction);
+      }
+
+      for (RecursiveAction task : tasks) {
+        getPool().execute(task);
+      }
+
+      QueuePusher<ResultBatch<T>> outputPusher = new QueuePusher<>(out, hasTimeout, timeoutAt);
+      List<BatchedResultsCursor<T>> intermediaryOutputsCursors = new ArrayList<>(intermediaryOutputs.size());
+      for (BlockingQueue<ResultBatch<T>> queue : intermediaryOutputs) {
+        intermediaryOutputsCursors.add(
+            new BlockingQueueuBatchedResultsCursor<>(queue, orderingFn, hasTimeout, timeoutAt)
+        );
+      }
+      PrepareMergeCombineInputsAction<T> finalMergeAction = new PrepareMergeCombineInputsAction<>(
+          intermediaryOutputsCursors,
+          outputPusher,
+          orderingFn,
+          combineFn,
+          yieldAfter,
+          batchSize,
+          cancellationGizmo
+      );
+
+      getPool().execute(finalMergeAction);
+    }
+
+    /**
+     * Computes maximum number of layer 1 parallel merging tasks given available processors and an estimate of current
+     * {@link ForkJoinPool} utilization. A return value of 1 or less indicates that a serial merge will be done on
+     * the pool instead.
+     */
+    int computeNumTasks()
+    {
+      final int availableProcessors = JvmUtils.getRuntimeInfo().getAvailableProcessors();
+      final int runningThreadCount = getPool().getRunningThreadCount();
+      final int submissionCount = getPool().getQueuedSubmissionCount();
+      // max is minimum of either number of processors or user suggested parallelism
+      final int maxParallelism = Math.min(availableProcessors, parallelism);
+      // adjust max to be no more than total pool parallelism less the number of running threads + submitted tasks
+      final int utilizationEstimate = runningThreadCount + submissionCount;
+      // minimum of 'max computed parallelism' and pool parallelism less current 'utilization estimate'
+      final int computedParallelism = Math.min(maxParallelism, getPool().getParallelism() - utilizationEstimate);
+      // compute total number of layer 1 'parallel' tasks, the final merge task will take the remaining slot
+      final int computedOptimalParallelism = Math.min(
+          (int) Math.floor((double) sequences.size() / 2.0),
```

Comment:
Would you add a comment where `2.0` came from?

Reply:
It came from needing at least 2 sequences per partition so there is something for the thread to merge.